attic/multibody/multibody_solvers.cc

```diff
@@ -0,0 +1,568 @@
+  prog_->SetSolverOption(SnoptSolver::id(), "Print file", filename_);
+  prog_->SetSolverOption(SnoptSolver::id(), "Major feasibility tolerance",
+                         major_tolerance_);
+  prog_->SetSolverOption(SnoptSolver::id(), "Minor feasibility tolerance",
+                         minor_tolerance_);
+
+  auto position_constraint = make_shared<PositionConstraint>(tree_);
+
+  prog_->AddConstraint(position_constraint, q_);
+
+  // Adding the fixed joint constraints
+  for (uint i = 0; i < fixed_joints.size(); i++) {
+    int ind = fixed_joints[i];
+    prog_->AddConstraint(q_(ind) == q(ind));
+  }
+
+  prog_->AddQuadraticCost((q_ - q).dot(q_ - q));
+
+  // The initial guess for q needs to be set up separately before calling Solve
+  program_result_ = drake::solvers::Solve(*prog_);
+
+  return program_result_;
+}
+
+bool PositionSolver::CheckConstraint(VectorXd q, double tolerance) const {
+  auto position_constraint = make_shared<PositionConstraint>(tree_);
+  bool check = position_constraint->CheckSatisfied(q, tolerance);
+  // If the constraint is not satisfied, print the status.
+  if (!check) {
+    std::cout << "Position constraint: "
+              << position_constraint->CheckSatisfied(q, tolerance) << std::endl;
+  }
+
+  return check;
+}
+
+shared_ptr<MathematicalProgram> PositionSolver::get_program() { return prog_; }
+
+MathematicalProgramResult PositionSolver::get_program_result() {
+  return program_result_;
+}
+
+VectorXd PositionSolver::GetSolutionQ() {
+  return program_result_.GetSolution(q_);
+}
+
+void PositionSolver::set_filename(string filename) { filename_ = filename; }
+
+void PositionSolver::set_major_tolerance(double major_tolerance) {
+  major_tolerance_ = major_tolerance;
+}
+void PositionSolver::set_minor_tolerance(double minor_tolerance) {
+  minor_tolerance_ = minor_tolerance;
+}
+
+string PositionSolver::get_filename() { return filename_; }
+
+double PositionSolver::get_major_tolerance() { return major_tolerance_; }
+
+double PositionSolver::get_minor_tolerance() { return minor_tolerance_; }
+
+ContactSolver::ContactSolver(const RigidBodyTree<double>& tree,
+                             ContactInfo contact_info)
+    : tree_(tree),
+      contact_info_(contact_info),
+      prog_(make_shared<MathematicalProgram>()) {
+  // Initializing the variable
+  q_ = prog_->NewContinuousVariables(tree_.get_num_positions(), "q");
+}
+
+void ContactSolver::SetInitialGuessQ(VectorXd q) {
+  prog_->SetInitialGuess(q_, q);
+}
+
+void ContactSolver::AddJointLimitConstraint(const double tolerance) {
+  VectorXd joint_min = tree_.joint_limit_min;
+  VectorXd joint_max = tree_.joint_limit_max;
+
+  for (int i = 0; i < joint_min.size(); ++i) {
+    // Adding minimum and maximum joint angle constraints.
+    prog_->AddConstraint(q_(i) >= (joint_min(i) + tolerance));
+    prog_->AddConstraint(q_(i) <= (joint_max(i) - tolerance));
+  }
+}
+
+MathematicalProgramResult ContactSolver::Solve(VectorXd q,
```

Comment:
Recommend renaming `Solve`, since it modifies the problem by fixing some decision variables. Alternatively, split into two functions.

Reply:
The `fixed_joints` parameter is optional though, so renaming it to reflect this could possibly be misleading. Do I go ahead and rename it?

Reply:
What do you think of splitting it into two? I just don't like the idea of the Solve method doing anything more than solving the program. It would make it hard, for instance, to re-run the problem.

Reply:
Even if it is split, we would end up with overlapping code. Can it be renamed to something like `FindSolution`?

Reply:
Ideally, we create optimization problems and then solve them, with the creation and the solving being two distinct and separable steps.

Reply:
I see. So something like `FormulateProblem` and then `Solve`?

Reply:
I'd recommend pulling out the setting of fixed joints into it's own method. The other parts related to setting up the `MathematicalProgram` could be done at construction of the `PositionSovler` (and other classes), correct?

Reply:
Yeah that can be done. We'll also need to provide the values of `q`, etc. (The ones used in the cost) to the constructor in that case. Or would it be better to do everything apart from setting the cost and solving in the constructor? The `Solve` function would then set up the cost and solve the program.

Reply:
If it makes sense, problem setup methods can be lumped together to simplify things from the user's perspective. The only thing I feel strongly about is separating out the solve from the setup.

Reply:
3c3fff9 tries to resolve this. I'm not sure if its the best way of doing it.